Multisig/UI/Safe Management/Create Safe/Web3AuthMFAService.swift

```diff
@@ -0,0 +1,208 @@
+        if !changeResult {
+            throw GSError.Web3AuthKeyReconstructionError(underlyingError: "Password not changed")
+        }
+    }
+
+    private func reconstructWithError() async throws {
+        do {
+            let _ = try await thresholdKey.reconstruct()
+        } catch {
+            throw GSError.Web3AuthKeyReconstructionError(description: "Failed to reconstruct key", underlyingError: error, howToFix: "More shares required.")
+        }
+    }
+
+    /// This is used on a new device after the user logged in with their social login and we detect, that we cannot
+    /// recreate the final key because we're missing the device share. We need to ask the user for their password
+    /// to recreate the device share and reconstruct the final key
+    func recreateDeviceShare(password: String? = nil) async throws {
+        if let password = password {
+            try await inputPassword(password)
+        }
+        try await reconstructWithError()
+        var generateShareStoreResult: GenerateShareStoreResult
+        do {
+            generateShareStoreResult = try await thresholdKey.generate_new_share()
+        } catch {
+            throw GSError.Web3AuthKeyReconstructionError(underlyingError: "Failed create new share")
+        }
+        try saveShare(shareIndex: generateShareStoreResult.hex)
+    }
+
+    private func saveShare(shareIndex: String) throws {
+        var share: String
+        do {
+            share = try thresholdKey.output_share(shareIndex: shareIndex, shareType: nil)
+
+        } catch {
+            throw GSError.Web3AuthKeyReconstructionError(underlyingError: "Failed to output share")
+        }
+        do {
+            let _ = try keychainService.save(data: Data(share.utf8), forKey: "\(publicAddress):device-key")
+        } catch {
+            throw GSError.Web3AuthKeyReconstructionError(underlyingError: "Failed to save share")
+        }
+    }
+
+    /// Instantiate a Web3AuthMFAService object.
+    /// If the postboxkey was not yet migrated it will be migrated to MFA.
+    /// Even if the flag `initialize` is not given. Unfortunately we cannot detect reliably if a key is SFA or MFA.
+    /// And calling ThresholdKey.initialize() will convert it.
+    ///
+    /// - Parameters:
+    ///   - postBoxKey: Key returned by CustomAuth
+    ///   - publicAddress: Used as a key to store the shares in the KeyChain
+    ///   - password: If given used as one share (optional)
+    ///   - initialize: If given, init() assumes it is called on a postBoxKey that is still on SFA.
+    ///
+    /// - Returns: `Web3AuthMFAService`
+    ///
+    /// - Throws: `RuntimeError`, indicates the setup failed. Password wrong, not enough shares to recover key, etc.
+    init(postBoxKey: String,
+         publicAddress: String,
+         password: String? = nil, // if not nil, use to create a password share on initialize otherwise used to unlock the key
```

Comment:
this comment would be better in the parameters description

Reply:
Done